### PR TITLE
trigram indexes

### PR DESCRIPTION
**Pull request by taroface**
DOC-4902
DOC-3818
DOC-3687
DOC-3686
DOC-3685

- Draft documentation for trigram indexes.
  - Updated the docs for session variables and operators where applicable.
- Made a few opportunistic small improvements to adjacent docs.

### CODE REVIEW

**Review by awoods187 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/15571)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dikshant, @mgartner, @rytaft, and @taroface)

___
*[`v22.2/create-index.md` line 141 at r1](https://reviewable.io/reviews/cockroachdb/docs/15571#-NGJp8CN2lPMBB8CFw4I:-NGJp8CN2lPMBB8CFw4J:b-vyy490) ([raw file](https://github.com/cockroachdb/docs/blob/0df35c39a85ebea763b310b6545ce350a320e367/v22.2/create-index.md#L141)):*
> ```Markdown
> ~~~
> 
> The preceding example is equivalent to the following PostgreSQL-compatible syntax:
> ```

This might not be the best example if you try it via demo:

CREATE INVERTED INDEX ON rides(city gin_trgm_ops);
ERROR: unimplemented: primary key column city cannot be present in an inverted index
SQLSTATE: 0A000
HINT: You have attempted to use a feature that is not yet implemented.
See: https://go.crdb.dev/issue-v/84405/v22.2

You can make it vehicle_city easily:
CREATE INVERTED INDEX ON rides(vehicle_city gin_trgm_ops);
___
*[`v22.2/create-index.md` line 145 at r1](https://reviewable.io/reviews/cockroachdb/docs/15571#-NGJri0MC75nQ25OvdKC:-NGJri0MC75nQ25OvdKD:b-f46ny) ([raw file](https://github.com/cockroachdb/docs/blob/0df35c39a85ebea763b310b6545ce350a320e367/v22.2/create-index.md#L145)):*
> ```Markdown
> {% include_cached copy-clipboard.html %}
> ~~~sql
> CREATE INDEX trgm_idx ON rides USING GIN (city gin_trgm_ops);
> ```

Why not show the PG example first since its what most people would be more familiar with?



_Code quote:_
```Markdown
CREATE INDEX trgm_idx ON rides USING GIN (city gin_trgm_ops);
```
___
*[`v22.2/trigram-indexes.md` line 9 at r1](https://reviewable.io/reviews/cockroachdb/docs/15571#-NGJsGdr5_FJ8cEYgoW6:-NGJsGdr5_FJ8cEYgoW7:bc2zd93) ([raw file](https://github.com/cockroachdb/docs/blob/0df35c39a85ebea763b310b6545ce350a320e367/v22.2/trigram-indexes.md#L9)):*
> ```Markdown
> ---
> 
> A _trigram index_ is a type of [inverted index](inverted-indexes.html) created on a [`STRING`](string.html) column. Trigram indexes are used to efficiently search for strings in large tables without providing an exact search term.
> ```

Same thing--why not call it GIN to use PG language?
___
*[`v22.2/trigram-indexes.md` line 19 at r1](https://reviewable.io/reviews/cockroachdb/docs/15571#-NGJsPmJBf8NGNFz_dFU:-NGJsPmJBf8NGNFz_dFV:bnqvlxk) ([raw file](https://github.com/cockroachdb/docs/blob/0df35c39a85ebea763b310b6545ce350a320e367/v22.2/trigram-indexes.md#L19)):*
> ```Markdown
> ## How do trigram indexes work?
> 
> Trigram indexes separate strings into trigrams. A trigram is a group of three consecutive characters in a string that is used to [measure the similarity of two strings](https://www.postgresql.org/docs/current/pgtrgm.html). 
> ```

Its technical three consecutive characters including spaces/blanks
___
*[`v22.2/trigram-indexes.md` line 21 at r1](https://reviewable.io/reviews/cockroachdb/docs/15571#-NGJsSam5-MyMitFLEsk:-NGJsSam5-MyMitFLEsl:b-kcigu3) ([raw file](https://github.com/cockroachdb/docs/blob/0df35c39a85ebea763b310b6545ce350a320e367/v22.2/trigram-indexes.md#L21)):*
> ```Markdown
> Trigram indexes separate strings into trigrams. A trigram is a group of three consecutive characters in a string that is used to [measure the similarity of two strings](https://www.postgresql.org/docs/current/pgtrgm.html). 
> 
> For example, use the `show_trgm()` [built-in function](functions-and-operators.html#trigrams-functions) to display the trigrams for a string (in this example, `word`):
> ```

all of the possible
___
*[`v22.2/trigram-indexes.md` line 72 at r1](https://reviewable.io/reviews/cockroachdb/docs/15571#-NGJsm0C9a6xECGbDaJo:-NGJsm0D9zYjS07qvYKU:bq50sa6) ([raw file](https://github.com/cockroachdb/docs/blob/0df35c39a85ebea763b310b6545ce350a320e367/v22.2/trigram-indexes.md#L72)):*
> ```Markdown
>   ~~~
> 
> - Using the PostgreSQL-compatible syntax:
> ```

See comment above on the ordering
___
*[`v22.2/trigram-indexes.md` line 115 at r1](https://reviewable.io/reviews/cockroachdb/docs/15571#-NGJswns3DP8aip-LRCK:-NGJswns3DP8aip-LRCL:b-mcod45) ([raw file](https://github.com/cockroachdb/docs/blob/0df35c39a85ebea763b310b6545ce350a320e367/v22.2/trigram-indexes.md#L115)):*
> ```Markdown
> ~~~
> 
> A partial index with trigram matching enabled:
> ```

Love this example of combining partial with trigram 
___
*[`v22.2/trigram-indexes.md` line 206 at r1](https://reviewable.io/reviews/cockroachdb/docs/15571#-NGJt6aACoIHfBDIbypN:-NGJt6aB4WdngMKJBeyr:b-iol2fb) ([raw file](https://github.com/cockroachdb/docs/blob/0df35c39a85ebea763b310b6545ce350a320e367/v22.2/trigram-indexes.md#L206)):*
> ```Markdown
> {% include_cached copy-clipboard.html %}
> ~~~ sql
> EXPLAIN SELECT w, similarity(w, 'word')
> ```

Nice use of explain too
___
*[`v22.2/trigram-indexes.md` line 335 at r1](https://reviewable.io/reviews/cockroachdb/docs/15571#-NGJtDbo09J6IMnC0DdB:-NGJtDbo09J6IMnC0DdC:bravt1y) ([raw file](https://github.com/cockroachdb/docs/blob/0df35c39a85ebea763b310b6545ce350a320e367/v22.2/trigram-indexes.md#L335)):*
> ```Markdown
> 
> ## Unsupported features
> ```

Link to github tracking issue?

**Review by taroface — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/15571)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @awoods187, @dikshant, @mgartner, and @rytaft)

___
*[`v22.2/create-index.md` line 141 at r1](https://reviewable.io/reviews/cockroachdb/docs/15571#-NGJp8CN2lPMBB8CFw4I:-NGSU-da7cJVW4cwd6Sj:b-26z2x4) ([raw file](https://github.com/cockroachdb/docs/blob/0df35c39a85ebea763b310b6545ce350a320e367/v22.2/create-index.md#L141)):*
<details><summary><i>Previously, awoods187 (Andy Woods) wrote…</i></summary><blockquote>

This might not be the best example if you try it via demo:

CREATE INVERTED INDEX ON rides(city gin_trgm_ops);
ERROR: unimplemented: primary key column city cannot be present in an inverted index
SQLSTATE: 0A000
HINT: You have attempted to use a feature that is not yet implemented.
See: https://go.crdb.dev/issue-v/84405/v22.2

You can make it vehicle_city easily:
CREATE INVERTED INDEX ON rides(vehicle_city gin_trgm_ops);

</blockquote></details>

Done, thanks for noticing this.
___
*[`v22.2/create-index.md` line 145 at r1](https://reviewable.io/reviews/cockroachdb/docs/15571#-NGJri0MC75nQ25OvdKC:-NGSU7U38XLtrb5oO_di:b-1qtpgz) ([raw file](https://github.com/cockroachdb/docs/blob/0df35c39a85ebea763b310b6545ce350a320e367/v22.2/create-index.md#L145)):*
<details><summary><i>Previously, awoods187 (Andy Woods) wrote…</i></summary><blockquote>

Why not show the PG example first since its what most people would be more familiar with?
</blockquote></details>

This was intended to match the existing GIN index example directly above this. Point well taken, though; I'll reorder the examples throughout.
___
*[`v22.2/trigram-indexes.md` line 9 at r1](https://reviewable.io/reviews/cockroachdb/docs/15571#-NGJsGdr5_FJ8cEYgoW6:-NGSs0QM-tznsZwFLVXT:b-yc5trm) ([raw file](https://github.com/cockroachdb/docs/blob/0df35c39a85ebea763b310b6545ce350a320e367/v22.2/trigram-indexes.md#L9)):*
<details><summary><i>Previously, awoods187 (Andy Woods) wrote…</i></summary><blockquote>

Same thing--why not call it GIN to use PG language?
</blockquote></details>

@awoods187 Would that preclude calling it a GiST index, which seems to be equivalent? I'm slightly confused on this point, because the actual inverted indexes doc doesn't mention GiST. https://www.cockroachlabs.com/docs/v22.1/inverted-indexes.html
___
*[`v22.2/trigram-indexes.md` line 19 at r1](https://reviewable.io/reviews/cockroachdb/docs/15571#-NGJsPmJBf8NGNFz_dFU:-NGSqO8N5Tl6WwD9aPk4:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/0df35c39a85ebea763b310b6545ce350a320e367/v22.2/trigram-indexes.md#L19)):*
<details><summary><i>Previously, awoods187 (Andy Woods) wrote…</i></summary><blockquote>

Its technical three consecutive characters including spaces/blanks
</blockquote></details>

Done.
___
*[`v22.2/trigram-indexes.md` line 21 at r1](https://reviewable.io/reviews/cockroachdb/docs/15571#-NGJsSam5-MyMitFLEsk:-NGSqR8A1_wT3tQb4TXj:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/0df35c39a85ebea763b310b6545ce350a320e367/v22.2/trigram-indexes.md#L21)):*
<details><summary><i>Previously, awoods187 (Andy Woods) wrote…</i></summary><blockquote>

all of the possible
</blockquote></details>

Done.
___
*[`v22.2/trigram-indexes.md` line 72 at r1](https://reviewable.io/reviews/cockroachdb/docs/15571#-NGJsm0C9a6xECGbDaJo:-NGSqZIjBH99N-6MAwG2:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/0df35c39a85ebea763b310b6545ce350a320e367/v22.2/trigram-indexes.md#L72)):*
<details><summary><i>Previously, awoods187 (Andy Woods) wrote…</i></summary><blockquote>

See comment above on the ordering
</blockquote></details>

Done.
___
*[`v22.2/trigram-indexes.md` line 115 at r1](https://reviewable.io/reviews/cockroachdb/docs/15571#-NGJswns3DP8aip-LRCK:-NGSqZoz7UC9oZmPhmyu:b-l3ag1v) ([raw file](https://github.com/cockroachdb/docs/blob/0df35c39a85ebea763b310b6545ce350a320e367/v22.2/trigram-indexes.md#L115)):*
<details><summary><i>Previously, awoods187 (Andy Woods) wrote…</i></summary><blockquote>

Love this example of combining partial with trigram 
</blockquote></details>

FYI, this is straight from Marcus's QA notes: https://gist.github.com/mgartner/62e43beb2c0de45ad787e495f4bc2171
___
*[`v22.2/trigram-indexes.md` line 335 at r1](https://reviewable.io/reviews/cockroachdb/docs/15571#-NGJtDbo09J6IMnC0DdB:-NGSrVX02XqVM0qspmep:bqmjwcz) ([raw file](https://github.com/cockroachdb/docs/blob/0df35c39a85ebea763b310b6545ce350a320e367/v22.2/trigram-indexes.md#L335)):*
<details><summary><i>Previously, awoods187 (Andy Woods) wrote…</i></summary><blockquote>

Link to github tracking issue?
</blockquote></details>

Added.

File: v22.2/inverted-indexes.md [hunk 1]
```diff
@@ -61,19 +61,19 @@ This lets you search based on subcomponents.
 
 You can use GIN indexes to improve the performance of queries using `JSONB` or `ARRAY` columns. You can create them:
 
-- At the same time as the table with the `INVERTED INDEX` clause of [`CREATE TABLE`](create-table.html#create-a-table-with-secondary-and-gin-indexes).
 - For existing tables with [`CREATE INDEX`](create-index.html):
```
Comment (kathancox) [inline v22.2/inverted-indexes.md:64]: At the end of the first bullet point here (after `CREATE INDEX`) there is a colon, assume that needs to change?

File: v22.2/inverted-indexes.md [hunk 2]
```diff
@@ -61,19 +61,19 @@ This lets you search based on subcomponents.
...
+- At the same time as the table with the `INVERTED INDEX` clause of [`CREATE TABLE`](create-table.html#create-a-table-with-secondary-and-gin-indexes).
```
Comment (kathancox) [inline v22.2/inverted-indexes.md:65]: nit: This feels a little abrupt here after `CREATE TABLE`. It might be nice to introduce the two sub-bullets with a couple of words. I assume these are one of two methods one would use depending on their scenario? Only something along the lines of " ... with either of the following relevant options/methods/?"

File: v22.2/trigram-indexes.md [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+title: Trigram Indexes
+summary: Trigram indexes on STRING columns improve the performance of fuzzy string matching and pattern matching in large datasets.
+toc: true
+keywords: gin, gin index, gin indexes, inverted index, inverted indexes, accelerated index, accelerated indexes, trigram, trigrams
+docs_area: develop
+---
+
+A _trigram index_ is a type of [inverted index](inverted-indexes.html) created on a [`STRING`](string.html) column. Trigram indexes are used to efficiently search for strings in large tables without providing an exact search term.
+
+This page describes to create and use trigram indexes on CockroachDB.
```
Comment (kathancox) [inline v22.2/trigram-indexes.md:11]: Typo: Missing "how" after "describes" here.

File: v22.2/trigram-indexes.md [hunk 4]
```diff
@@ -0,0 +1,354 @@
+---
+title: Trigram Indexes
+summary: Trigram indexes on STRING columns improve the performance of fuzzy string matching and pattern matching in large datasets.
+toc: true
+keywords: gin, gin index, gin indexes, inverted index, inverted indexes, accelerated index, accelerated indexes, trigram, trigrams
+docs_area: develop
+---
+
+A _trigram index_ is a type of [inverted index](inverted-indexes.html) created on a [`STRING`](string.html) column. Trigram indexes are used to efficiently search for strings in large tables without providing an exact search term.
```
Comment (kathancox) [inline v22.2/trigram-indexes.md:9]: Is this new to v22.2? Maybe add the "New in" tag to the start of this paragraph?

File: v22.2/trigram-indexes.md [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+- **equality**: [`=`](functions-and-operators.html#operators). Note that standard [`btree` secondary indexes](create-index.html#parameters) may perform better than trigram indexes for equality searches.
+- **pattern matching (case-sensitive)**: [`LIKE`](functions-and-operators.html#operators)
+- **pattern matching (case-insensitive)**: [`ILIKE`](functions-and-operators.html#operators)
+- **similarity matching**: [`%`](functions-and-operators.html#operators). This operator returns `true` if the strings in the comparison have a similarity that meets or exceeds the threshold set by the `pg_trgm.similarity_threshold` [session variable](show-vars.html#pg_trgm_similarity_threshold).
+
+For usage examples, see [Use a trigram index to speed up fuzzy string matching](#use-a-trigram-index-to-speed-up-fuzzy-string-matching).
+
+## Examples
+
+### Simple examples
```
Comment (kathancox) [inline v22.2/trigram-indexes.md:99]: In my opinion, we might want to stay away from using the word "Simple" especially for examples... as some users may not find it easy. Perhaps the title for this subsection could also be more descriptive like the second one? "Create various trigram indexes" "Create indexes with different trigram matching"... Not sure, if they are helpful 😬 But, some ideas!

File: v22.2/trigram-indexes.md [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+  ('norm'),
+  ('weird'),
+  ('worried'),
+  ('wofoord'),
+  ('wobarrd');
+
+INSERT INTO t SELECT 'empty' FROM generate_series(1, 10000);
+~~~
+
+Retrieve the columns with values similar to `word`, using the `%` operator. Sort the results by the output of the `similarity()` [built-in function](functions-and-operators.html#trigrams-functions):
```
Comment (kathancox) [inline v22.2/trigram-indexes.md:163]: nit: somewhere before running the `SELECT` query **without** the trigram function it might be clearer to mention that. As in, "First, retrieve the columns with values similar.... without using a trigram index." I feel like this will help reinforce that we're kind of evaluating the difference in performance between the two to begin.

File: v22.2/trigram-indexes.md [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+SELECT similarity('weird', 'word');
+~~~
+
+~~~
+      similarity
+----------------------
+  0.2222222222222222
+~~~
+
+Notice that the fuzzy search took 30 milliseconds to execute. Without a trigram index, the statement performs a full scan:
```
Comment (kathancox) [inline v22.2/trigram-indexes.md:202]: nit: Could be good to refer to the `EXPLAIN` docs page here. "Without a trigram index, the statement performs a full scan. Check this with `EXPLAIN`...: " Or, something along those lines?

File: v22.2/trigram-indexes.md [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+      │
+      └── • scan
+            estimated row count: 0 (<0.01% of the table; stats collected 15 minutes ago; using stats forecast for 42 minutes in the future)
+            table: t@t_w_idx
+            spans: 1 span
+~~~
+
+## Unsupported features
+
+The following PostgreSQL syntax and features are currently unsupported. For details, see the [tracking issue](https://github.com/cockroachdb/cockroach/issues/41285).
```
Comment (kathancox) [inline v22.2/trigram-indexes.md:336]: Does this need to be added to the known limitations page?

**Review by kathancox — APPROVED** [review]
LGTM, couple of optional comments and one typo!

File: v22.2/trigram-indexes.md (hunk 5 above)
Comment (taroface) [reply v22.2/trigram-indexes.md:99]: Agree!

File: v22.2/trigram-indexes.md (hunk 6 above)
Comment (taroface) [reply v22.2/trigram-indexes.md:163]: Great point; I clarified this.

**Review by awoods187 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/15571)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dikshant, @jordanlewis, @kathancox, @mgartner, @rytaft, and @taroface)

___
*[`v22.2/create-index.md` line 145 at r1](https://reviewable.io/reviews/cockroachdb/docs/15571#-NGJri0MC75nQ25OvdKC:-NGsGVX860Umzj04fq9T:b-ucfknh) ([raw file](https://github.com/cockroachdb/docs/blob/0df35c39a85ebea763b310b6545ce350a320e367/v22.2/create-index.md#L145)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

This was intended to match the existing GIN index example directly above this. Point well taken, though; I'll reorder the examples throughout.
</blockquote></details>

Thanks!
___
*[`v22.2/trigram-indexes.md` line 9 at r1](https://reviewable.io/reviews/cockroachdb/docs/15571#-NGJsGdr5_FJ8cEYgoW6:-NGsG_3i44lDQnFHBOIZ:b-wr3o9f) ([raw file](https://github.com/cockroachdb/docs/blob/0df35c39a85ebea763b310b6545ce350a320e367/v22.2/trigram-indexes.md#L9)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

@awoods187 Would that preclude calling it a GiST index, which seems to be equivalent? I'm slightly confused on this point, because the actual inverted indexes doc doesn't mention GiST. https://www.cockroachlabs.com/docs/v22.1/inverted-indexes.html
</blockquote></details>

@jordanlewis or @rytaft can you weigh in here?

File: v22.2/trigram-indexes.md (hunk 8 above)
Comment (taroface) [reply v22.2/trigram-indexes.md:336]: Adding per @lnhsingh

**Comment by taroface** [comment]
@jordanlewis TFTR! I revised accordingly; PTAL.

Where you mentioned "inexact for a similar_to search", does this mean `%`? We don't have `similar_to` in our docs. Let me know if it needs to be added.

**Comment by taroface** [comment]
@jordanlewis TFTR! I revised accordingly; PTAL.

Where you mentioned "inexact for a similar_to search", does this mean `%`? We don't have `similar_to` in our docs. Let me know if it needs to be added.

**Comment by jordanlewis** [comment]
> Where you mentioned "inexact for a similar_to search", does this mean %? We don't have similar_to in our docs. Let me know if it needs to be added.

The `%` operator is the same as comparing the result of the `similarity()` operator (sorry, there is no such thing as `similar_to`, I had it wrong) to the `pg_trgm.similarity_threshold`.

**Comment by taroface** [comment]
> > Where you mentioned "inexact for a similar_to search", does this mean %? We don't have similar_to in our docs. Let me know if it needs to be added.
> 
> The `%` operator is the same as comparing the result of the `similarity()` operator (sorry, there is no such thing as `similar_to`, I had it wrong) to the `pg_trgm.similarity_threshold`.

Sounds good; that was my assumption. Unless you had any other comments on the updated PR, I will merge this!

File: v22.2/trigram-indexes.md [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+
+This page describes how to create and use trigram indexes on CockroachDB.
+
+{{site.data.alerts.callout_info}}
+Some PostgreSQL syntax and features are currently unsupported. For details, see [Unsupported features](#unsupported-features).
+{{site.data.alerts.end}}
+
+## How do trigram indexes work?
+
+Trigram indexes make [substring and similarity matches](https://www.postgresql.org/docs/current/pgtrgm.html) efficient by indexing the unique trigrams of a string. A trigram is a group of three consecutive characters (including spaces) in a string.
```
Comment (jordanlewis) [inline v22.2/trigram-indexes.md:19]: Just noticed that there's another slight flaw here - actually, trigrams don't include spaces or whitespace. It looks like they do, because the trigrams at the beginning and end of a word (words are produced by splitting the input on whitespace, and discarding the whitespace) include artificially inserted spaces to help distinguish the starts and ends of words.

tl;dr You may not need to include this detail but trigrams don't really include the whitespace in strings.

File: v22.2/trigram-indexes.md [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+
+To display all of the possible trigrams for a string, use the `show_trgm()` [built-in function](functions-and-operators.html#trigrams-functions):
```
Comment (jordanlewis) [inline v22.2/trigram-indexes.md:21]: nit: it's not the possible trigrams, it's all of the trigrams. Maybe say "to display the trigrams within a string"?

**Review by jordanlewis — APPROVED** [review]
LGTM besides the last two comments, feel free to fix and merge! Thanks.

File: v22.2/trigram-indexes.md (hunk 9 above)
Comment (taroface) [reply v22.2/trigram-indexes.md:19]: Thanks for correcting this! Fixed.
